modules/core/src/main/scala/doobie/util/fragments.scala

```diff
@@ -5,68 +5,103 @@
 package doobie
 package util
 
-import doobie.implicits._
-import cats.Reducible
+import cats.data.NonEmptyList
 import cats.syntax.all._
+import cats.{Foldable, Reducible}
+import doobie.implicits._
 
 /** Module of `Fragment` constructors. */
 object fragments {
 
-  /** Returns `VALUES (fs0), (fs1), ...`. */
+  /** Returns `VALUES fs0, fs1, ...`. */
   def values[F[_]: Reducible, A](fs: F[A])(implicit w: util.Write[A]): Fragment =
-    fs.toList.map(a => fr0"(${w.toFragment(a)})").foldSmash1(fr0"VALUES ", fr",", fr"")
-
-  /** Returns `f IN (fs0, fs1, ...)`. */
-  def in[F[_]: Reducible, A: util.Put](f: Fragment, fs: F[A]): Fragment =
-    fs.toList.map(a => fr0"$a").foldSmash1(f ++ fr0"IN (", fr",", fr")")
-
-  /** Returns `f IN ((fs0-A, fs0-B), (fs1-A, fs1-B), ...)`. */
-  def in[F[_]: Reducible, A: util.Put, B: util.Put](f: Fragment, fs: F[(A,B)]): Fragment =
-    fs.toList.map { case (a,b) => fr0"($a,$b)" }.foldSmash1(f ++ fr0"IN (", fr",", fr")")
-
-  /** Returns `f NOT IN (fs0, fs1, ...)`. */
-  def notIn[F[_]: Reducible, A: util.Put](f: Fragment, fs: F[A]): Fragment =
-    fs.toList.map(a => fr0"$a").foldSmash1(f ++ fr0"NOT IN (", fr",", fr")")
-
-  /** Returns `(f1) AND (f2) AND ... (fn)`. */
-  def and(fs: Fragment*): Fragment =
-    fs.toList.map(parentheses).intercalate(fr"AND")
-
-  /** Returns `(f1) AND (f2) AND ... (fn)` for all defined fragments. */
+    fs.toList.map(a => fr0"${w.toFragment(a)}").foldSmash1(fr0"VALUES ", fr",", fr"")
+
+  /** Returns `(f IN (fs0, fs1, ...))`. */
+  def in[A: util.Put](f: Fragment, fs0: A, fs1: A, fs: A*): Fragment =
+    in(f, fs0 :: fs1 :: fs.toList)
+
+  /** Returns `(f IN (fs0, fs1, ...))`, or `false` for empty `fs`. */
+  def in[F[_]: Foldable, A: util.Put](f: Fragment, fs: F[A]): Fragment = fs.toList.toNel match {
+    case Some(fs) => parentheses(f ++ fr"IN" ++ parentheses(comma(fs.map(a => fr"$a"))))
+    case None => fr"${false}"
+  }
+
+  /** Returns `(f IN ((fs0-A, fs0-B), (fs1-A, fs1-B), ...))`, or `false` for empty `fs`. */
+  def in[F[_]: Foldable, A: util.Put, B: util.Put](f: Fragment, fs: F[(A,B)]): Fragment = fs.toList.toNel match {
+    case Some(fs) => parentheses(f ++ fr"IN" ++ parentheses(comma(fs.map { case (a,b) => fr0"($a,$b)" })))
+    case None => fr"${false}"
+  }
+
+  /** Returns `(f NOT IN (fs0, fs1, ...))`. */
+  def notIn[A: util.Put](f: Fragment, fs0: A, fs1: A, fs: A*): Fragment =
+    notIn(f, fs0 :: fs1 :: fs.toList)
+
+  /** Returns `(f NOT IN (fs0, fs1, ...))`, or `true` for empty `fs`. */
+  def notIn[F[_]: Foldable, A: util.Put](f: Fragment, fs: F[A]): Fragment = fs.toList.toNel match {
+    case Some(fs) => parentheses(f ++ fr"NOT IN" ++ parentheses(comma(fs.map(a => fr"$a"))))
+    case None => fr"${true}"
+  }
+
+  /** Returns `(f1 AND f2 AND ... fn)`. */
+  def and(f1: Fragment, f2: Fragment, fs: Fragment*): Fragment =
```

Comment:
Nit: I'm not too sure whether we need these 2+ param var args. If you statically have more than two fragments you can directly interpolate it in right?

```
fr"...WHERE $frag1 AND $frag2 AND $frag3"
```

which IMO is also more readable than


```
fr"...WHERE ${and(frag1, frag2, frag3)}"
```

Reply:
Well, they were here before. But I can remove them, if you want :+1:

Reply:
True you're right.